data/crac/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/cnec/AbstractBranchCnec.java

```diff
@@ -0,0 +1,177 @@
+
+    protected AbstractBranchCnec(String id, String name, NetworkElement networkElement, State state, boolean optimized,
+                              boolean monitored, Set<BranchThreshold> thresholds) {
+        super(id, name, networkElement, state, optimized, monitored);
+        setThresholds(thresholds);
+    }
+
+    protected AbstractBranchCnec(String id, NetworkElement networkElement, State state, boolean optimized,
+                              boolean monitored, Set<BranchThreshold> thresholds) {
+        super(id, networkElement, state, optimized, monitored);
+        setThresholds(thresholds);
+    }
+
+    @Override
+    public double computeMargin(double actualValue, Side side, Unit unit) {
+        double marginOnLowerBound = actualValue - getLowerBound(side, unit).orElse(Double.NEGATIVE_INFINITY);
+        double marginOnUpperBound = getUpperBound(side, unit).orElse(Double.POSITIVE_INFINITY) - actualValue;
+        return Math.min(marginOnLowerBound, marginOnUpperBound);
+    }
+
+    @Override
+    public final Set<BranchThreshold> getThresholds() {
+        return thresholds;
+    }
+
+    @Override
+    public void addThreshold(BranchThreshold branchThreshold) {
+        checkAndInitThreshold(branchThreshold);
+        bounds.resetBounds();
+        thresholds.add(branchThreshold);
+    }
+
+    public void setThresholds(Set<BranchThreshold> thresholds) {
+        thresholds.forEach(this::checkAndInitThreshold);
+        bounds.resetBounds();
+        this.thresholds = new HashSet<>(thresholds);
+    }
+
+    /**
+     * This method aims to define the side of thresholds when it does not require network so that margins and min/max
+     * values can be computed without synchronization.
+     * We check first if the threshold is compatible with the cnec physical parameter and then we set the side
+     * if it is possible.
+     *
+     * @param branchThreshold: Threshold to check and update if possible.
+     */
+    private void checkAndInitThreshold(BranchThreshold branchThreshold) {
+        checkThreshold(branchThreshold);
+        switch (branchThreshold.getRule()) {
+            case ON_LEFT_SIDE:
+            case ON_REGULATED_SIDE:
+                branchThreshold.setSide(Side.LEFT);
+                break;
+            case ON_RIGHT_SIDE:
+            case ON_NON_REGULATED_SIDE:
+                branchThreshold.setSide(Side.RIGHT);
+                break;
+            default:
+                break;
+        }
+    }
+
+    @Override
+    public void synchronize(Network network) {
+        Branch<?> branch = checkAndGetValidBranch(network, networkElement.getId());
+        setVoltageLevel(Side.LEFT, branch.getTerminal1().getVoltageLevel().getNominalV());
+        setVoltageLevel(Side.RIGHT, branch.getTerminal2().getVoltageLevel().getNominalV());
+        thresholds.forEach(threshold -> {
+            switch (threshold.getRule()) {
+                case ON_LEFT_SIDE:
+                case ON_REGULATED_SIDE:
```

Comment:
the regulated side is always left or is it ucte specific ? 

if it is ucte specific, maybe add a todo here.